src/Decompiler/Analysis/TrashedRegisterFinder.cs

```diff
@@ -459,6 +463,126 @@ private bool ProcedureTerminates(Expression expr)
             return (p != null && flow[p].TerminatesProcess);
         }
 
+        /// <summary>
+        /// Backpropagate stack pointer from procedure return.
+        /// Assume that stack pointer at the end of procedure has the same
+        /// value as at the start
+        /// </summary>
+        /// <example>
+        /// If we have
+        /// <code>
+        ///     call eax; We do not know calling convention of this indirect call
+        ///             ; So we do not know value of stack pointer after it
+        /// cleanup:
+        ///     pop esi
+        ///     pop ebp
+        ///     ret
+        /// </code>
+        /// then we could assume than stack pointer at "cleanup" label is
+        /// "fp - 8"
+        /// </example>
+        // $REVIEW: It is highly unlikely that there is a procedure that
+        // leaves the stack pointer at different values depending on what
+        // path you took through it. Should we encounter such procedures in
+        // a binary we might consider turning this analysis off with a user
+        // switch.
+        private void BackpropagateStackPointer()
+        {
+            foreach (var proc in procedures)
+            {
+                foreach (var block in proc.ExitBlock.Pred)
+                {
+                    BackpropagateStackPointer(block);
+                }
+            }
+        }
+
+        private Expression GetValue(SymbolicEvaluationContext ctx, Storage stg)
+        {
+            if (!ctx.RegisterState.TryGetValue(stg, out var value))
+                return Constant.Invalid;
+            return value;
+        }
+
+        private void SetValue(
+            SymbolicEvaluationContext ctx,
+            Storage stg,
+            Expression value)
+        {
+            ctx.RegisterState[stg] = value;
+        }
+
+        private void ClearEvaluationState(Block block)
```

Comment:
The naming had me a little confused here. Consider using `CreateEvaluationState` instead to clarify the intent.

Reply:
Done (rebased)